index.js

```diff
@@ -1,174 +1,175 @@
-        opts.code = opts.codec
-        opts.codec = multicodec.get(opts.code).name
-      }
-      if (typeof opts.source === 'undefined' &&
-          typeof opts.data === 'undefined') {
-        throw new Error('Block instances must be created with either an encode source or data')
-      }
-      if (typeof opts.source !== 'undefined' && !opts.codec && !opts.code) {
-        throw new Error('Block instances created from source objects must include desired codec')
-      }
-      if (opts.data && !opts.cid && !opts.codec && !opts.code) {
-        throw new Error('Block instances created from data must include cid or codec')
-      }
-      if (!opts.cid && !opts.algo) opts.algo = 'sha2-256'
-      // Do our best to avoid accidental mutations of the options object after instantiation
-      // Note: we can't actually freeze the object because we mutate it once per property later
-      opts = Object.assign({}, opts)
-      Object.defineProperty(this, 'opts', readonly(opts))
-    }
+const { coerce, isBinary } = bytes
+const copyBinary = value => {
+  const b = coerce(value)
+  return coerce(b.buffer.slice(b.byteOffset, b.byteOffset + b.byteLength))
+}
 
-    source () {
-      if (this.opts.cid || this.opts.data ||
-          this._encoded || this._decoded) return null
-      if (!this.opts.source) return null
-      return this.opts.source
-    }
+const clone = obj => transform(obj, (result, value, key) => {
+  const cid = CID.asCID(value)
+  if (cid) {
+    result[key] = cid
+  } else if (isBinary(value)) {
+    result[key] = copyBinary(value)
+  } else if (typeof value === 'object' && value !== null) {
+    result[key] = clone(value)
+  } else {
+    result[key] = value
+  }
+})
 
-    async cid () {
-      if (this.opts.cid) return this.opts.cid
-      const hash = await multihash.hash(this.encodeUnsafe(), this.opts.algo)
-      const cid = CID.create(1, this.code, hash)
-      this.opts.cid = cid
-      // https://github.com/bcoe/c8/issues/135
-      /* c8 ignore next */
-      return cid
-    }
+const setImmutable = (obj, key, value) => {
+  if (typeof value === 'undefined') throw new Error(`${key} cannot be undefined`)
+  Object.defineProperty(this, key, readonly(value))
+}
 
-    get codec () {
-      if (this.opts.code) {
-        this.opts.codec = multicodec.get(this.opts.code).name
-      }
-      if (this.opts.cid) {
-        this.opts.codec = multicodec.get(this.opts.cid.code).name
-      }
-      return this.opts.codec
-    }
+class Block {
+  constructor ({ codec, hasher, source, cid, data }) {
+    setImmutable(this, '_codec', codec)
+    setImmutable(this, '_hasher', hasher)
+    if (source) setImmutable(this, '_source', source)
+    if (cid) setImmutable(this, '_cid', cid)
+    if (data) setImmutable(this, '_data', data)
+    if (!source && (!data || !cid)) throw new Error('Missing required argument')
+    if (source && (!codec || !hasher)) throw new Error('Missing required argument')
```

Comment:
I think things would be a lot cleaner if `Block.encoder` and `Block.decoder` would do above invariant checks instead of deferring that to shared constructor.

Reply:
agreed, i’m going to move all of these in the new patch to `js-multiformats`.